include/seqan3/alignment/cigar_conversion/alignment_from_cigar.hpp

```diff
@@ -0,0 +1,226 @@
+        {
+            reference_length += cigar_count;
+            query_length += cigar_count;
+        }
+        else if ('D'_cigar_operation == cigar_operation)
+        {
+            reference_length += cigar_count;
+        }
+        else if ('I'_cigar_operation == cigar_operation)
+        {
+            query_length += cigar_count;
+        }
+    }
+
+    if (static_cast<size_t>(zero_based_reference_start_position + reference_length) > std::ranges::size(reference))
+        throw std::logic_error{"The CIGAR string indicates a reference length of at least "
+                               + std::to_string(zero_based_reference_start_position + reference_length)
+                               + " but the supplied reference sequence is only of size"
+                               + std::to_string(std::ranges::size(reference))};
+
+    // Get soft clipping at the start and end of the cigar string
+    // -------------------------------------------------------------------------
+    uint32_t soft_clipping_start{0};
+    uint32_t soft_clipping_end{0};
+
+    // Checks if the given index in the cigar vector is a soft clip.
+    auto soft_clipping_at = [&](size_t const index)
+    {
+        return cigar_vector[index] == 'S'_cigar_operation;
+    };
+    // Checks if the given index in the cigar vector is a hard clip.
+    auto hard_clipping_at = [&](size_t const index)
+    {
+        return cigar_vector[index] == 'H'_cigar_operation;
+    };
+    // Checks if the given cigar vector as at least min_size many elements.
+    auto vector_size_at_least = [&](size_t const min_size)
+    {
+        return cigar_vector.size() >= min_size;
+    };
+    // Returns the cigar count of the ith cigar element in the given cigar vector.
+    auto cigar_count_at = [&](size_t const index)
+    {
+        return get<0>(cigar_vector[index]);
+    };
+
+    // check for soft clipping at the first two positions
+    // cigar is non-empty, checked at the very beginning.
+    if (soft_clipping_at(0))
+        soft_clipping_start = cigar_count_at(0);
+    else if (vector_size_at_least(2) && hard_clipping_at(0) && soft_clipping_at(1))
+        soft_clipping_start = cigar_count_at(1);
+
+    // Check for soft clipping at the last two positions. But only if the vector size has at least 2, respectively
+    // 3 elements. Accordingly, if the following arithmetics overflow they are protected by the corresponding
+    // if expressions below.
+    auto last_index = cigar_vector.size() - 1;
+    auto second_last_index = last_index - 1;
+
+    if (vector_size_at_least(2) && soft_clipping_at(last_index))
+        soft_clipping_end = cigar_count_at(last_index);
+    else if (vector_size_at_least(3) && hard_clipping_at(last_index) && soft_clipping_at(second_last_index))
+        soft_clipping_end = cigar_count_at(second_last_index);
+
+    if (soft_clipping_start + query_length + soft_clipping_end != std::ranges::size(query))
```

Comment:
The only reason we compute `soft_clipping_end` is for this check?

Reply:
yep.. it might cause serious trouble when assigning the sequence to the gap decoratpr of the alignment and computing it shouldn't affect the performance too much.